woocommerce/woocommerce
# add screenshot utility

### Changes proposed in this Pull Request:

When a timeout occurs in an E2E test, `jest` closes the browser context before triggering an error. Then there isn't a browser context available to screenshot the error. 

This PR adds error handling to trap errors where there isn't a browser context for a screenshot. It refactors the screenshot code to add a `takeScreenshotFor` utility which could be used for debugging tests where a timeout is occurring.

### How to test the changes in this Pull Request:

1. Run E2E tests twice without re-initializing the container
```
WC_E2E_SCREENSHOTS=1 npx wc-e2e test:e2e
```
2. The second run should create screenshots for the setup tests that fail
3. Create a test `test-screenshot.js` in the specs folder
```
/* eslint-disable jest/no-export, jest/no-disabled-tests */
const { merchant } = require( '@woocommerce/e2e-utils' );
const { takeScreenshotFor } = require( '@woocommerce/e2e-environment' );
const {
	it,
	describe,
	beforeAll,
} = require( '@jest/globals' );

	describe('Store owner can login and make sure WooCommerce is activated', () => {
		beforeAll(async () => {
			await merchant.login();
		});

		it('can make sure WooCommerce is activated. If not, activate it', async () => {
			await merchant.openPlugins();
			await takeScreenshotFor( 'test plugin page' );
		});

	});
```
4. Run the test
```
npx wc-e2e test:e2e tests/e2e/specs/test-screenshot.js
```
5. A `test-plugin-page.png` should be created and be a screenshot of the plugin page.

### Changelog entry

N/A
